Add support for reporting metrics to SignalFx

rohitvyavahare opened the pull request.

Commit:
SignalFx support for kafka monitoring

Commit:
change config

Commit:
add signalfx-java dep

mpetazzoni commented on config/kafka-monitor.properties:
```diff
@@ -46,7 +46,10 @@
     "bootstrap.servers": "localhost:9092",
+
     "produce.record.delay.ms": 100,
+    "produce.treat.zero.throughput.as.unavailable": false,
```
Is that a good default? Any particular reason why we want this, if it wasn't the default in kafka-monitor to begin with?

rohitvyavahare replied:
@mpetazzoni  here is the description of that config : 
https://github.com/linkedin/kafka-monitor/blob/master/src/main/java/com/linkedin/kmf/services/configs/ProduceServiceConfig.java#L67

But yes I think, we should not introduce this here, we can add this in our run.py so that it will be signalfx specific.

mpetazzoni commented on config/kafka-monitor.properties:
```diff
@@ -77,62 +80,15 @@
 
-  "reporter-service": {
-    "class.name": "com.linkedin.kmf.services.DefaultMetricsReporterService",
+  "signalfx-service": {
```
If we intend to contribute this upstream, I don't think you can mess around with the default configuration this much. Maybe provide an example of how to configure the `signalfx-service` in the class Javadoc of `SignalFxMetricsReporterService`?

rohitvyavahare replied:
yes make sense

mpetazzoni commented on config/kafka-monitor.properties:
```diff
@@ -77,62 +80,15 @@
-      "kmf.services:type=produce-service,name=*:produce-error-rate",
-      "kmf.services:type=consume-service,name=*:consume-error-rate"
-    ]
+    "metric.dimenstions": [
```
typo

mpetazzoni commented on config/kafka-monitor.properties:
```diff
@@ -77,62 +80,15 @@
+    "metric.dimenstions": [
+    ],
+    "report.signalfx.url": "",
+    "token" :""
```
spacing

mpetazzoni commented on src/main/java/com/linkedin/kmf/services/SignalFxMetricsReporterService.java:
```diff
@@ -0,0 +1,197 @@
+/**
+ * Copyright 2016 LinkedIn Corp. Licensed under the Apache License, Version 2.0 (the "License"); you may not use this
```
That's not correct. I don't think LinkedIn would own the copyright for this file. Check with the contribution guidelines of the project.

rohitvyavahare replied:
@mpetazzoni  Yes, I tried to change it, but it has checkstyle config which checks for this config 
https://github.com/signalfx/kafka-monitor/blob/master/checkstyle/checkstyle.xml#L21. What's your thought on this?

mpetazzoni replied:
You should check what their contributor agreement is, and whether they require that they own the copyright, etc.

rohitvyavahare replied:
@mpetazzoni @ramjothikumar thinks we should involve legal team here.

mpetazzoni commented on src/main/java/com/linkedin/kmf/services/SignalFxMetricsReporterService.java:
```diff
@@ -0,0 +1,197 @@
+public class SignalFxMetricsReporterService implements Service {
+  private static final Logger LOG = LoggerFactory.getLogger(SignalFxMetricsReporterService.class);
+
+  private final String _name;
```
Is the "underscore for private variables" a convention in this project, or is this just yours?

rohitvyavahare replied:
@mpetazzoni yes, a convention in this project

mpetazzoni reviewed (changes requested):
Still reviewing, but some initial feedback

Commit:
add example of how to configure ignalfx in config and fix typo in SignalFxMetricsReporterServiceConfig.java

Commit:
keep default values of zookeeper and bootstrap

mpetazzoni commented on config/kafka-monitor.properties:
```diff
@@ -96,42 +96,52 @@
-#      "kmf.services:type=consume-service,name=*:consume-error-rate"
-#    ]
-#  }
+  #  Example statsd-service to report metrics
```
Looks like you have an extra space before the `#`; remove so those don't appear as they've changed

mpetazzoni commented on src/main/java/com/linkedin/kmf/services/configs/SignalFxMetricsReporterServiceConfig.java:
```diff
@@ -0,0 +1,69 @@
+  public static final String SIGNALFX_METRIC_DIMENSION = "metric.dimensions";
+  public static final String SIGNALFX_METRIC_DIMENSION_DOC = "Dimensions added to each metric. Example: [\"key1:value1\", \"key2:value2\"] ";
+  
+  public static final String SIGNALFX_TOKEN = "token";
```
For consistency, you might want to name this `report.signalfx.token`

mpetazzoni commented on src/main/java/com/linkedin/kmf/services/configs/SignalFxMetricsReporterServiceConfig.java:
```diff
@@ -0,0 +1,69 @@
+  public static final String REPORT_SIGNALFX_URL = "report.signalfx.url";
+  public static final String REPORT_SIGNALFX_URL_DOC = "The url of signalfx server which SignalFxMetricsReporterService will report the metrics values.";
+
+  public static final String SIGNALFX_METRIC_DIMENSION = "metric.dimensions";
```
For consistency, you might want to name this `report.metrics.dimensions`

mpetazzoni commented on src/main/java/com/linkedin/kmf/services/configs/SignalFxMetricsReporterServiceConfig.java:
```diff
@@ -0,0 +1,69 @@
+import org.apache.kafka.common.config.AbstractConfig;
+import org.apache.kafka.common.config.ConfigDef;
+
+public class SignalFxMetricsReporterServiceConfig extends AbstractConfig {
```
Add Javadoc

mpetazzoni commented on src/main/java/com/linkedin/kmf/services/SignalFxMetricsReporterService.java:
```diff
@@ -0,0 +1,197 @@
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+
```
Remove empty line here

mpetazzoni commented on src/main/java/com/linkedin/kmf/services/SignalFxMetricsReporterService.java:
```diff
@@ -0,0 +1,197 @@
+  private Map<String, String> _dimensionsMap;
+  
+  public SignalFxMetricsReporterService(Map<String, Object> props, String name) throws Exception {
+      
```
Remove empty line here

mpetazzoni commented on src/main/java/com/linkedin/kmf/services/SignalFxMetricsReporterService.java:
```diff
@@ -0,0 +1,197 @@
+    _signalfxUrl = config.getString(SignalFxMetricsReporterServiceConfig.REPORT_SIGNALFX_URL);
+    _signalfxToken = config.getString(SignalFxMetricsReporterServiceConfig.SIGNALFX_TOKEN);
+    
+    if (_signalfxToken.length() < 1) {
```
Use `StringUtils.isNullOrEmpty`

mpetazzoni commented on src/main/java/com/linkedin/kmf/services/SignalFxMetricsReporterService.java:
```diff
@@ -0,0 +1,197 @@
+    _metricRegistry = new MetricRegistry();
+    _metricMap = new HashMap<String, SettableDoubleGauge>();
+    _dimensionsMap = new HashMap<String, String>();
+    setUpDimensionMap();
```
It's usually better design to have something like this:

```java
_dimensionsMap = setUpDimensionsMap();
```

And have the method create the map with the correct contents, instead of just filling it (and depending on the fact that you've correctly initialized the map beforehand)

mpetazzoni commented on src/main/java/com/linkedin/kmf/services/SignalFxMetricsReporterService.java:
```diff
@@ -0,0 +1,197 @@
+        _metricRegistry,
+        _signalfxToken
+    );
+    if (_signalfxUrl.length() > 1) {
```
It should probably be an error if the report URL isn't set

rohitvyavahare replied:
@mpetazzoni isn't it default to prod url? I want to give this option for our internal use.

mpetazzoni commented on src/main/java/com/linkedin/kmf/services/SignalFxMetricsReporterService.java:
```diff
@@ -0,0 +1,197 @@
+    }
+    
+    _metricMetadata = _signalfxReporter.getMetricMetadata();    
+    _signalfxReporter.start(_reportIntervalSec, TimeUnit.SECONDS);    
```
This should probably happen in `start()`

mpetazzoni commented on src/main/java/com/linkedin/kmf/services/SignalFxMetricsReporterService.java:
```diff
@@ -0,0 +1,197 @@
+  @Override
+  public synchronized void stop() {
+    _executor.shutdown();
+    LOG.info("{}/SignalFxMetricsReporterService stopped", _name);
```
You also need to stop the `_signalfxReporter` here, otherwise you can't guarantee that everything's been sent correctly before shutting down.

mpetazzoni commented on src/main/java/com/linkedin/kmf/services/SignalFxMetricsReporterService.java:
```diff
@@ -0,0 +1,197 @@
+  @Override
+  public void awaitShutdown() {
+    try {
+      _executor.awaitTermination(Integer.MAX_VALUE, TimeUnit.MILLISECONDS);
```
Don't use an infinite timeout

mpetazzoni commented on src/main/java/com/linkedin/kmf/services/SignalFxMetricsReporterService.java:
```diff
@@ -0,0 +1,197 @@
+    }
+    for (String dimension : _dimensions) {
+      String dimensionKey = dimension.split(":")[0];
+      String dimensionValue = dimension.split(":")[1];
```
Don't split twice:

```java
String[] parts = dimensions.split(":");
_dimensionsMap.put(parts[0], parts[1]);
```

mpetazzoni commented on src/main/java/com/linkedin/kmf/services/SignalFxMetricsReporterService.java:
```diff
@@ -0,0 +1,197 @@
+    String scheme = url.getProtocol();
+    String host = url.getHost();
+    int port = url.getPort();
+    return new SignalFxEndpoint(scheme, host, port);    
```
Local variables here aren't particularly interesting, you can just do:

```java
return new SignalFxEndpoint(url.getProtocol(), url.getHost(), url.getPort());
```

mpetazzoni commented on src/main/java/com/linkedin/kmf/services/SignalFxMetricsReporterService.java:
```diff
@@ -0,0 +1,197 @@
+        if (parts.length < 2 || !parts[1].contains("cluster-monitor")) {
+          continue;
+        }
+        String signalFxMetricName = generateSignalFxMetricName(attributeValue.mbean(), attributeValue.attribute());        
```
It's a bit sad that you have to recompute this every time when the set of metrics doesn't change. Can you maintain a map of `MbeanAttributeValue` to `SettableDoubleGauge`?

Commit:
avoid recalculating metric name and change metric name to maintain format

mpetazzoni commented on src/main/java/com/linkedin/kmf/services/SignalFxMetricsReporterService.java:
```diff
@@ -164,34 +160,35 @@ private void captureMetrics() {
-    _metricMap.get(metric).setValue(value);
+  private void setMetricValue(MbeanAttributeValue attributeValue) {
+    String key = attributeValue.mbean() + attributeValue.attribute();
+    if (!_metricMap.containsKey(key)) {
```
Use `computeIfAbsent`:

```java
Metric metric = _metricMap.computeIfAbsent(key, n -> createMetric(attributeValue));
metric.setValue(attributeValue.value());
```

mpetazzoni commented on src/main/java/com/linkedin/kmf/services/SignalFxMetricsReporterService.java:
```diff
@@ -155,7 +151,7 @@ private void captureMetrics() {
 
       for (final MbeanAttributeValue attributeValue : attributeValues) {
-        String metric = attributeValue.toString();
+        String metric = attributeValue.toString();        
```
Remove trailing whitespace.

Also, all this code (lines 154-162) is kinda weird. What's the "shape" of what you're working from, and what are you trying to get to?

mpetazzoni commented on src/main/java/com/linkedin/kmf/services/SignalFxMetricsReporterService.java:
```diff
@@ -164,34 +160,35 @@ private void captureMetrics() {
   
-  private void createMetric(String metric) {
-    LOG.info("Creating metric : " + metric);
+  private void createMetric(MbeanAttributeValue attributeValue) {
```
I think this method should return the `Metric` object that was created, and have whoever calls it insert it into the `_metricMap` (which will happen with the `computeIfAbsent()` call I've suggested above).

mpetazzoni commented on src/main/java/com/linkedin/kmf/services/SignalFxMetricsReporterService.java:
```diff
@@ -164,34 +160,35 @@ private void captureMetrics() {
     }
-    if (metric.contains("partition")) {          
-      String partitionNumber = "" + metric.charAt(metric.length() - 1);
+    if (signalFxMetricName.contains("partition")) {          
```
If you're extracting the partition number as a dimension, you should probably also remove it from the metric name?

mpetazzoni commented on src/main/java/com/linkedin/kmf/services/SignalFxMetricsReporterService.java:
```diff
@@ -164,34 +160,35 @@ private void captureMetrics() {
     SettableDoubleGauge s = _metricMetadata.forMetric(new SettableDoubleGauge())
-        .withMetricName(metric).metric();    
+        .withMetricName(signalFxMetricName).metric();    
     for (String key : _dimensionsMap.keySet()) {
```
Use `.entrySet()` instead so do you don't do a hash lookup on every iteration

Commit:
accept dimensions as json

Commit:
add computeIfAbsent and support for java 1.8 in gradle

rohitvyavahare commented on build.gradle:
```diff
@@ -6,7 +6,7 @@ allprojects {
   version = "1.0.0"
 
-  sourceCompatibility = 1.7
+  sourceCompatibility = 1.8
```
@mpetazzoni to support java 1.8 we need to change this. If we want to make a pull request then it might create compatibility issue, what's your thought?

mpetazzoni replied:
Unless you're using Java 8 specific features in SignalFxMetricsReporterService (which you really don't need to), you shouldn't have to change this. The fact that we execute this code with a Java8 JVM has no impact here (JVMs are backwards compatible).

rohitvyavahare replied:
@mpetazzoni Yes I am using java 8 specific, as you suggested to use 
```java
_metricMap.computeIfAbsent(key, n -> createMetric(attributeValue));
```
It give me this error if I use 1.7 
```shell
/Users/rohitvyavahare/work/kafka-monitor/src/main/java/com/linkedin/kmf/services/SignalFxMetricsReporterService.java:158: error: lambda expressions are not supported in -source 1.7
    SettableDoubleGauge metric = _metricMap.computeIfAbsent(key, n -> createMetric(attributeValue));
                                                                   ^
  (use -source 8 or higher to enable lambda expressions)
```

Commit:
keep partition as metric name instead of partition-number

1 - add partition metric logic
2 - add javadoc to signalfxconfig
3 - remove unnecessary step for creating _dimensionsMap
4 - format config file

Commit:
remove extra comma from config

Commit:
start signalfxreporter in start function

Commit:
Change log format

mpetazzoni commented on src/main/java/com/linkedin/kmf/services/SignalFxMetricsReporterService.java:
```diff
@@ -0,0 +1,188 @@
+  private Map<String, String> _dimensionsMap;
+
+  public SignalFxMetricsReporterService(Map<String, Object> props, String name) throws Exception {
+
```
Remove empty line

mpetazzoni commented on src/main/java/com/linkedin/kmf/services/SignalFxMetricsReporterService.java:
```diff
@@ -0,0 +1,188 @@
+    }
+
+    SignalFxReporter.Builder sfxReportBuilder = new SignalFxReporter.Builder(
+        _metricRegistry,
```
You probably don't need to wrap this (and the two lines below)

mpetazzoni commented on src/main/java/com/linkedin/kmf/services/SignalFxMetricsReporterService.java:
```diff
@@ -0,0 +1,188 @@
+        );
+    if (_signalfxUrl.length() > 1) {
+      SignalFxEndpoint signalFxEndpoint = getSignalFxEndpoint(_signalfxUrl);
+      _signalfxReporter = sfxReportBuilder.setEndpoint(signalFxEndpoint).build();
```
Remove the `.build()` and the else to just do a `.build()` in either case:

```java
if (!StringUtils.isEmpty(_signalfxUrl)) {
  sfxReportBuilder.setEndpoint(getSignalFxEndpoint(_signalfxUrl));
}
_signalfxReporter = sfxReportBuilder.build();
```

mpetazzoni commented on src/main/java/com/linkedin/kmf/services/SignalFxMetricsReporterService.java:
```diff
@@ -0,0 +1,188 @@
+    String service = bean.split(":")[1];
+    String serviceType = service.split(",")[1].split("=")[1];
+    String[] segs = {serviceType, attribute};
+    return StringUtils.join(segs, ".");
```
Why not simply:

```java
return String.format("%s.%s", serviceType, attribute);
```

mpetazzoni commented on src/main/java/com/linkedin/kmf/services/SignalFxMetricsReporterService.java:
```diff
@@ -0,0 +1,188 @@
+  private void captureMetrics() {
+    for (String metricName : _metricNames) {
+      String mbeanExpr = metricName.substring(0, metricName.lastIndexOf(":"));
+      String attributeExpr = metricName.substring(metricName.lastIndexOf(":") + 1);
```
Cache `metricName.lastIndexOf(':')` into a variable so you don't have to do the work twice. Also use the variant that takes a `character` instead of a `String` (it's faster)

mpetazzoni reviewed (changes requested):
Your commit message also needs improvement. It's pretty clear that it's for kafka-monitor, so maybe something like this is clearer:

> Add support for reporting metrics to SignalFx

Commit:
SignalfxReporterService clean up

1 - remove empty line
2 - wrap up _signalfxReporter creation
3 - wrap up generating metric name
4 - cache metricName index

Commit:
Change lambda to java1.7 style
